lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,551 @@
+	    lookahead_cr_lf(parser->cursor) ||
+	    lookahead_crlf(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Don't tokenize response. */
+	if (is_resp(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Stop if there is a new line to parse. */
+	if (lookahead_cr_lf_and_more(parser->cursor) ||
+	    lookahead_crlf_and_more(parser->cursor)) {
+		return -EAGAIN;
+	}
+
+	if (parser->count == 0) {
+		*token = at_match_cmd(parser->cursor, &remainder);
+	} else {
+		*token = at_match_subparam(parser->cursor, &remainder);
+	}
+
+	if (token->type == AT_TOKEN_TYPE_INVALID) {
+		/* If the token is neither a valid command nor a valid subparameter, check if it's a
+		 * non-quoted string.
+		 */
+		*token = at_match_str(parser->cursor, &remainder);
+		if (token->type == AT_TOKEN_TYPE_INVALID) {
+			return -EBADMSG;
+		}
+	}
+
+	if (is_subparam(token)) {
+		/* if the token is a subparameter but does not have a trailing comma then it is the
+		 * last subparameter of the line.
+		 * If there is more to parse, then the token must be followed by CRLF, CR, or LF,
+		 * otherwise the string is malformed.
+		 */
+		bool lookahead = lookahead_cr_lf_or_more(remainder) ||
+				 lookahead_crlf_or_more(remainder);
+
+		if (token->variant != AT_TOKEN_VARIANT_COMMA) {
+			if (remainder[0] != NULL_TERMINATOR && !lookahead) {
+				return -EBADMSG;
+			}
+		} else {
+			/* The next token is of type empty if and only if the current token has a
+			 * trailing comma and the remaining string either is empty or starts with
+			 * CRLF, CR, or LF.
+			 */
+			if (remainder[0] == NULL_TERMINATOR || lookahead) {
+				parser->is_next_empty = true;
+			}
+		}
+	}
+
+finalize:
+	parser->count++;
+	parser->cursor = remainder;
+
+	return 0;
+}
+
+/* Seek the AT parser cursor to the given index. */
+static int at_parser_seek(struct at_parser *parser, size_t index, struct at_token *token)
+{
+	int err;
+
+	if (!is_index_ahead(parser, index)) {
+		/* Rewind parser. */
+		parser->cursor = parser->at;
+		parser->count = 0;
+	}
+
+	do {
+		err = at_parser_tok(parser, token);
+	} while (is_index_ahead(parser, index) && !err);
+
+	return err;
+}
+
+int at_parser_init(struct at_parser *parser, const char *at)
+{
+	if (!parser || !at) {
+		return -EINVAL;
+	}
+
+	memset(parser, 0, sizeof(struct at_parser));
+
+	parser->at = at;
+	parser->cursor = at;
+	parser->init_sentinel = INIT_SENTINEL;
+
+	return 0;
+}
+
+int at_parser_next(struct at_parser *parser)
+{
+	int err;
+	struct at_token token = {0};
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	do {
+		err = at_parser_tok(parser, &token);
+	} while (!err);
+
+	if (err != -EAGAIN) {
+		return -EOPNOTSUPP;
+	}
+
+	trim_crlf(&parser->cursor);
+
+	/* Reset count. */
+	parser->count = 0;
+	/* Set pointer of current AT command string to the current cursor, which points to the
+	 * beginning of a new AT command line.
+	 */
+	parser->at = parser->cursor;
+
+	return 0;
+}
+
+int at_parser_cmd_get(struct at_parser *parser, const char **at)
```

Comment:
What's the use case for this function? Was there a case where `at_parser` was initialized somewhere, and the parsing happened somewhere else?

Reply:
I removed it. At some point it was replaced with a different function where it was needed but I forgot to remove this one.